docs/API/V1/README.md

```diff
@@ -0,0 +1,86 @@
+
+## Accessing dataset inventory
+
+The Pbench agent code uploads a dataset to the Pbench server in the form of a tarball with a carefully designed format. That tarball contains a hierarchical file system which the Pbench server uses to index and expose details collected by the agent during the course of a pbench benchmark run.
+
+While there are many ways for a client to access the metrics and metadata associated with a dataset, a client can also directly access the dataset tarball inventory.
+
+The [contents](contents.md) API exposes the directory hierarchy within the tarball, from the root directory `/` through the leaf files.
+
+The [inventory](inventory.md) API returns the full byte stream of any regular file within the directory hierarchy, including log files, postprocessed JSON files, and benchmark result text files.
+
+Starting from the root directory of the dataset tarball, the "contents" API allows the client to discover the entire file system structure, including all directories, regular files, and symbolic links.
+
+The "inventory" API allows the client to retrieve any regular file for further analysis.
+
+### Example
+
+```
+    def directory(request, level: int, name: str = "/", url: str):
+        ls = request.get(url).get_json()
+        print(f"{'  '*level}{name}")
+        for d in ls.directories:
+            directory(request, level + 1, d.name, d.url)
+        for f in ls.files:
+            print(f"{'  '*(level+1)}{f.name})
+            bytes = request.get(f.url)
+            # display byte stream:
+            # inline on terminal doesn't really make sense
+
+    directory(request, 0, "/", "http://host.example.com/api/v1/contents/<dataset>/")
+```
+
+__What do we do with a symlink?__
+
+> It could, in theory, be a link to either a file or a directory within the tarball. In which case, the URL for that item would differ, and the client can't blindly `GET` that URL without understanding the context. This breaks the simple recursive descent algorithm I tried to model.
+>
+> One potential solution would be to put a symlink to a regular file in the `files` list and a symlink to a directory in the `directories` list. Note that the symlink "file object" includes the `"linkpath"` of a symlink; however the API will have to figure out whether that's a directory or a regular file, either with another Elasticsearch query or via a `FileTree` check.
+>
+>__NOTE__: that the most common symlink in current archives seems to be a `reference-result` link to the `sample1` directory in the same parent directory, which would be easy to check.
+>
+>__ALTERNATIVE__: symlinks to directories would duplicate a directory tree in a straightforward contents traversal; e.g., both `reference-result` and `sample1` trees, which are identical. Another option is to treat a symlink specially, and not produce a URL; the target will either appear elsewhere in the tree and be handled normally, or it's a broken symlink that can't be followed anyway.
```

Comment:
The "file object" currently defined in Elasticsearch, and preserved in the `contents` API, does identify a file as a symlink. The big question would be how we manage moving directory symlinks into the `directories` list, which is currently just a list of strings. I'm already proposing to expand that to an object at least containing the string and URL; however to make these into a full "file object" like the `files` list we'd need to come up with the `mtime`/`size`/`mode` data from a filesystem since that information's not indexed. Or maybe we ignore those and just go with `name`, `url`, and `type` which would be (for a "directory") either `sym` or `dir`.

Reply:
I'm not proposing lying about symlinks in the lists (or even about omitting them from the lists).  I just think we should minimize the impact of them (in terms of ease of use as well as documentation).  That is, the URL that we return for them should access their referent, the directory or the file, so that if the client hits it, it will act very much like hitting a symlink in a filesystem.  Nevertheless, I don't think we should hide them completely -- I think the client should, if it _really_ cares, be able to tell whether a given entry is a symlink or not.  As I said elsewhere, I doubt that this will matter much in the case of files (although, we would have to do something for broken links), but it could be critical for directories, to allow clients to avoid infinite file walks due to symlink-produced loops.

Reply:
> I want to add the reference URL, and pointing a symlink at the `contents` API to get the file object again doesn't make much sense. (And leaves the caller floundering to figure out the proper URL.)

I'm not sure what you're saying here....  (In no case should we leave the caller floundering.)  For a symlink, I think we should return either the `contents` or `inventory` URL for the referent, depending on whether the referent is a directory or a file, respectively.  For a broken link, I would return either one -- I'm assuming that it'll generate a `NOT_FOUND`, either way.  If we've already reported the contents of the symlink in the file object, then the caller can already deal with the link if they want to.  And as long as the caller can consult the `type` before following the URL, I think everything is covered.